### ISSUE TITLE
Recursive user state data

### ISSUE THREAD

**Issue by Whiteknight** [Whiteknight/ParserObjects#29]
We need some kind of state recursion mechanism, so the user can push state on entry to a rule, and pop the state again when the rule exits (successfull or unsuccessfully). Ideally we would have a cascading mechanism, so if a value isn't found in current state, it can be retrieved from a prior state.

Would also like a mechanism so we can insert a value before a parse, and remove that value again after the parse. Would like guaranteed setup/cleanup of state before and after a parser is invoked. This includes the need to use the output of a parser as data to store at each point. 

These requirements suggest we want data with more structure, such as a <string, object> dictionary so users can retrieve values by name instead of by callback.

In summary:
1. Recursive, cascading state data, probably dictionary-based
1. Be able to setup state before, and cleanup state after, a parse
1. Be able to use the output of a parser as state (probably contextual) and be able to use a value from state during parser operation.

**Comment by Whiteknight** [comment]
Everything except the second half of #3 is in the v3 branch right now. There is a CreateParser which can be used to create a parser at parse time using (among other things) current state data. I'll have to explore a few more options before I declare this ticket done, I'm not entirely happy with the current implementation or that final limitation.

**Comment by Whiteknight** [comment]
I have made several cleanups to this code in v3 and I am happier with it now. I am marking this ticket closed.
